balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,627 @@
+// to the next address/subconn in our list.  We assume that NewSubConn will never
+// return an error.
+func (b *pickfirstBalancer) requestConnection() {
+	if !b.addressIndex.isValid() || b.state == connectivity.Shutdown {
+		return
+	}
+	curAddr, err := b.addressIndex.currentAddress()
+	if err != nil {
+		// This should not never happen because we already check for validity and
+		// return early above.
+		return
+	}
+	sd, ok := b.subConns.Get(*curAddr)
+	if !ok {
+		sd, err = newScData(b, *curAddr)
+		if err != nil {
+			// This should never happen.
+			b.logger.Warningf("Failed to create a subConn for address %v: %v", curAddr.String(), err)
+			b.state = connectivity.TransientFailure
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("error creating connection: %v", err)},
+			})
+			b.addressIndex.reset()
+			return
+		}
+		b.subConns.Set(*curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressIndex.increment() {
+			b.endFirstPass()
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Warningf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, state balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSd, found := b.subConns.Get(*sd.addr); !found || activeSd != sd {
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if state.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		b.addressIndex.seekTo(sd.addr)
+		b.state = connectivity.Ready
+		b.firstErr = nil
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, shutdown and re-connect when
+	// prompted.
+	if state.ConnectivityState == connectivity.Idle && b.state == connectivity.Ready {
+		b.state = connectivity.Idle
+		b.addressIndex.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker: &idlePicker{
+				exitIdle: b.ExitIdle,
+			},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			if b.state == connectivity.Idle {
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			if b.firstErr == nil {
+				b.firstErr = state.ConnectionError
+			}
```

Comment:
Does it make more sense to store the last connection error instead of the first one because that might be more meaningful since it is more recent?

Reply:
I feel that the last error might be more helpful because it is more recent. Also, if you think it would be better to do it as a follow-up PR, I'm fine with that too. But, if you think its easy to fold it in here, that works too. Up to you. Or we can start a thread on the chat group to see what Doug thinks.

Reply:
Changed to use the last error now.

Reply:
Thanks.